compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -6112,7 +6120,25 @@ private BType checkFieldAccessExpr(BLangFieldBasedAccess fieldAccessExpr, BType
             if (fieldAccessExpr.fieldKind == FieldKind.WITH_NS) {
                 resolveXMLNamespace((BLangFieldBasedAccess.BLangNSPrefixedFieldBasedAccess) fieldAccessExpr);
             }
-            BType laxFieldAccessType = getLaxFieldAccessType(varRefType);
+
+            BType laxFieldAccessType;
+            if (fieldAccessExpr.isWithCheckExpr && fieldAccessExpr.expectedType.tag != TypeTags.NONE) {
```

Comment:
Instead of doing this here, can we not let field access be type-checked normally, and instead do these checks and set the final type in `visitCheckAndCheckPanicExpr`?

Reply:
That way we may be able to remove the `isWithCheckExpr` field too. Doesn't feel right to be adding that info to `BLangFieldBasedAccess` anyway.

Reply:
Changed the logic related to this. (removed the `isWithCheckExpr` flag)